plugins/dmxusb/src/dmxusbopenrx.cpp

```diff
@@ -0,0 +1,338 @@
+        {
+            // Last frame was shorter, just put the current value
+            emit valueChanged(UINT_MAX, m_inputBaseLine, i - 2, current_payload[i]);
+            qDebug() << "Channel" << i - 2 << "changed to" << QString::number((uchar) current_payload[i], 10);
+        }
+    }
+}
+
+void DMXUSBOpenRx::run()
+{
+    // Wait for device to settle in case the device was opened just recently
+    // Also measure, whether timer granularity is OK
+    QTime time;
+    time.start();
+    usleep(1000);
+    if (time.elapsed() > 3)
+        m_granularity = Bad;
+    else
+        m_granularity = Good;
+
+    if (interface()->type() == DMXInterface::QtSerial)
+    {
+        if (DMXUSBWidget::open(0) == false)
+        {
+            close(0);
+            return;
+        }
+
+        if (interface()->clearRts() == false)
+        {
+            close(0);
+            return;
+        }
+    }
+
+    m_running = true;
+
+    QByteArray payload;
+    QByteArray& last_payload = m_inputLines[0].m_compareData;
+    QByteArray& current_payload = m_inputLines[0].m_universeData;
+
+    quint32 missed_frames = 0;
+    quint32 erroneous_frames = 0;
+    quint32 erroneous_reads = 0;
+
+    m_frameTimeUs = 0;
+
+    while (m_running == true)
+    {
+        payload = interface()->read(1024);
+
+        if (payload.length() == 0)
+        {
+            usleep(1000); // nothing to read, don't waste CPU
+            missed_frames += 1;
+        } 
+        else if (payload.length() == 1)
+        {
+            // a new frame has begun, the chip returns us the first byte
+            current_payload.append(payload);
+            usleep(500); // wait a little for the other bytes to be read
+        } 
+        else
+        {
+            current_payload.append(payload);
+
+            if (payload.length() > 600)
+            {
+                // we are not reading only a single frame but several of them in one chunk.
+                erroneous_reads += 1;
+                current_payload.clear();
+                qDebug() << interface()->serial() << "Erroneous read" << payload.length() << "bytes";
+
+                if (erroneous_reads > 10)
+                {
+                    // set low latency in order to try to fetch frames one by one.
+                    interface()->setLowLatency(true);
+                    erroneous_reads = 0;
+                }
+                continue;
+            }
+
+            if (current_payload.length() != last_payload.length() && erroneous_frames < 5)
+            {
+                qDebug() << interface()->serial() << "Bogus frame" << current_payload.length() << "bytes instead of" << last_payload.length();
+                current_payload.clear();
+                erroneous_frames += 1;
+                continue;
+            }
+
+            // a frame has been received
+            if (missed_frames > 300) // only to emit the debug message once, not at each frame
```

Comment:
Are you sure this logic is correct ?
missed_frames is reset to 0 two lines below. To me it looks like missed_frames will never reach 300.
Also 
- maybe "missed_frames" is not a proper name since the comment is "a frame has been received"
- "300", "600" and other magic numbers should be placed as defines at the beginning of the cpp file so we can track what they are and easily change them if needed

Reply:
Yes, I'm pretty sure of the logic. But it needs more explanation:
- when we reach the comment "a frame has been received", it means that all the checks have succedded, so the current_payload contains a valid frame
- but if missed_frames is > 300, it means that there was more than 300 empty reads in the previous cycles
- so we are receiving a correct frame after a long break (probably because the attached controller was plugged or switched on)
- so we display this fact on the debug logger. The whole purpose of this condition is to display on the console the fact that the widget now receives frames, without cluttering the console with a message on each frame received.

You can see that counters have been reset just after that.

I'll add this as a comment at this place to clarify this thing.

Reply:
For the magic numbers, you are right.

I'll move them, and explain their semantics.